main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -559,7 +560,139 @@ object Lowering {
+      }
+
+      def mkVarExp(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation): Expression = {
+        Expression.Var(sym, tpe, loc)
+      }
+
+      def mkWait(chan: Expression): Expression = {
+        Expression.GetChannel(chan, chan.tpe, Type.Impure, chan.eff, chan.loc.asSynthetic)
+      }
+
+      def mkLetSym(prefix: String, e: Expression): Symbol.VarSym = {
+        val name = prefix + Flix.Delimiter + flix.genSym.freshId()
+        Symbol.freshVarSym(name, BoundBy.Let, e.loc.asSynthetic)
+      }
+
+      def mkLetCont(prefix: String, e: Expression): (Symbol.VarSym, Expression => Expression) = {
+        val sym = mkLetSym(prefix, e)
+        val loc = e.loc.asSynthetic
+        (sym, e1 =>
+          Expression.Let(sym,
+            Modifiers(List(Ast.Modifier.Synthetic)),
+            e,
+            e1,
+            e1.tpe,
+            Type.mkAnd(e.pur, e1.pur, loc),
+            Type.mkUnion(e.eff, e1.eff, loc),
+            loc))
+      }
+
+      def mkStmCont(e: Expression): Expression => Expression = {
+        val loc = e.loc.asSynthetic
+        e1: Expression =>
+          Expression.Stm(
+            e,
+            e1,
+            e1.tpe,
+            Type.mkAnd(e.pur, e1.pur, loc),
+            Type.mkUnion(e.pur, e1.pur, loc),
+            loc
+          )
+      }
+
+      val funLoc = loc1.asSynthetic
+      val parLoc = loc0.asSynthetic
+      val fun = visitExp(exp)
+      val args = visitExps(exps)
+
+      // Create channels
+      val (funChanSym, funChan) = mkLetCont("chan", mkChannel(fun))
+      val chanStart = (List(funChanSym), funChan)
+      val (chanSyms, chans) = args.foldLeft(chanStart) {
+        case ((syms, bindings), e1) =>
+          val (s, c) = mkLetCont("chan", mkChannel(e1))
+          (syms ::: List(s), e => bindings(c(e)))
+      }
+
+      // Spawn expressions into channels
+      val spawns = chanSyms.tail.zip(args).foldLeft(mkStmCont(mkSpawn(funChanSym, fun))) {
+        case (spawnCont, (chan, e)) =>
+          e1 => spawnCont(mkStmCont(mkSpawn(chan, e))(e1))
+      }
+
+      // Wait for message in channels
+      val (funWaitSym, funWaitCont) = mkLetCont("tmp", mkWait(mkVarExp(funChanSym, fun.tpe, fun.loc)))
+      val (waitSyms, waits) = chanSyms.tail.zip(args).foldLeft((List(funWaitSym), funWaitCont)) {
+        case ((syms, letWaitCont), (chanSym, e0)) =>
+          val (s, l) = mkLetCont("tmp", mkWait(mkVarExp(chanSym, e0.tpe, e0.loc)))
+          (syms ::: List(s), (e: Expression) => letWaitCont(l(e)))
+      }
+
+      val fun1 = mkVarExp(funWaitSym, fun.tpe, funLoc)
+      val args1 = waitSyms.zip(args).map {
+        case (sym, e) => mkVarExp(sym, e.tpe, e.loc.asSynthetic)
+      }
+      val allPurs = Type.mkAnd(fun.tpe :: args.map(_.tpe), funLoc)
+      val allEffs = Type.mkUnion(fun.eff :: args.map(_.eff), funLoc)
+      val apply = Expression.Apply(fun1, args1, tpe, pur, eff, funLoc)
```

Comment:
You need to create applies for each `args1` here because the function is curried. Thats like the problem.

Reply:
Try to test with a function that takes 1 argument.

Reply:
Should I also return a curried Apply expression?